miniapps/multidomain/multidomain.cpp

```diff
@@ -0,0 +1,249 @@
+#include "mfem.hpp"
+#include <fstream>
+#include <iostream>
+
+using namespace mfem;
+
+void velocity_profile(const Vector &c, Vector &q)
+{
+   double A = 1.0;
+   double x = c(0);
+   double y = c(1);
+   double r = sqrt(pow(x, 2.0) + pow(y, 2.0));
+
+   q(0) = 0.0;
+   q(1) = 0.0;
+
+   if (abs(r) >= 0.25 - 1e-8)
+   {
+      q(2) = 0.0;
+   }
+   else
+   {
+      q(2) = A * exp(-(pow(x, 2.0) / 2.0 + pow(y, 2.0) / 2.0));
+   }
+}
+
+class ConvectionDiffusionTDO : public TimeDependentOperator
+{
+public:
+   ConvectionDiffusionTDO(ParFiniteElementSpace &h1_fes,
+                          Array<int> ess_tdofs,
+                          double alpha = 1.0,
+                          double kappa = 1.0e-1)
+      : TimeDependentOperator(h1_fes.GetTrueVSize()),
+        Mform(&h1_fes),
+        Kform(&h1_fes),
+        ess_tdofs_(ess_tdofs),
+        M_solver(h1_fes.GetComm()),
+        T_solver(h1_fes.GetComm())
+   {
+      Mform.AddDomainIntegrator(new MassIntegrator);
+      Mform.Assemble(0);
+      Mform.FormSystemMatrix(ess_tdofs_, M);
+
+      q = new VectorFunctionCoefficient(h1_fes.GetParMesh()->Dimension(),
+                                        velocity_profile);
+
+      Kform.AddDomainIntegrator(new ConvectionIntegrator(*q, -alpha));
+
+      d = new ConstantCoefficient(-kappa);
+      Kform.AddDomainIntegrator(new DiffusionIntegrator(*d));
+      Kform.Assemble(0);
+      Array<int> empty;
+      Kform.FormSystemMatrix(empty, K);
+
+      M_solver.iterative_mode = false;
+      M_solver.SetRelTol(1e-8);
+      M_solver.SetAbsTol(0.0);
+      M_solver.SetMaxIter(100);
+      M_solver.SetPrintLevel(0);
+      M_prec.SetType(HypreSmoother::Jacobi);
+      M_solver.SetPreconditioner(M_prec);
+      M_solver.SetOperator(M);
+
+      t1.SetSize(height);
+      t2.SetSize(height);
+   }
+
+   void Mult(const Vector &u, Vector &du_dt) const override
+   {
+      K.Mult(u, t1);
+      M_solver.Mult(t1, du_dt);
+      du_dt.SetSubVector(ess_tdofs_, 0.0);
+   }
+
+   ~ConvectionDiffusionTDO()
+   {
+      delete q;
+      delete d;
+   }
+
+   ParBilinearForm Mform, Kform;
+   HypreParMatrix M, K, C, *T = nullptr;
+   VectorCoefficient *q;
+   Coefficient *d;
+   Array<int> ess_tdofs_;
+
+   double current_dt = -1.0;
+
+   CGSolver M_solver;
+   HypreSmoother M_prec;
+
+   CGSolver T_solver;
+   HypreSmoother T_prec;
+
+   mutable Vector t1, t2;
+};
+
+int main(int argc, char *argv[])
```

Comment:
Segmentation fault for >4 procs.

```
mpirun -np 5 multidomain                                                                      ✔  5s  
Mesh::GeneratePartitioning(...): edgecut = 54


Assertion failed: (i>=0 && i<size) is false:
 --> Access element -1 of array, size = 11
 ... in function: const T &mfem::Array<mfem::Element *>::operator[](int) const [T = mfem::Element *]
 ... in file: mesh/../general/array.hpp:746

Backtrace:


Assertion failed: (i>=0 && i<size) is false:
 --> Access element -1 of array, size = 11
 ... in function: const T &mfem::Array<mfem::Element *>::operator[](int) const [T = mfem::Element *]
 ... in file: mesh/../general/array.hpp:746

Backtrace:
0) [0x556077f40ae7]: mfem::mfem_error(char const*)
1) [0x556078074429]: mfem::Array<mfem::Element*>::operator[](int) const
2) [0x5560780752d1]: mfem::Mesh::GetFaceGeometry(int) const
3) [0x5560783f81df]: mfem::FiniteElementSpace::GetFaceDofs(int, mfem::Array<int>&, int) const
4) [0x556078475a04]: mfem::ParFiniteElementSpace::GetFaceDofs(int, mfem::Array<int>&, int) const
5) [0x5560783fecd9]: mfem::FiniteElementSpace::GetFaceVDofs(int, mfem::Array<int>&) const
6) [0x5560781c5a60]: mfem::ParSubMesh::Transfer(mfem::ParGridFunction const&, mfem::ParGridFunction&)
7) [0x556077f12c5c]: main
8) [0x7fe08cc1830f]: __libc_start_call_main
9) [0x7fe08cc183c0]: __libc_start_main
10) [0x556077f11af4]: _start

Lookup backtrace source lines:
addr2line -C -e multidomain 0x556077f40ae7 0x556078074429 0x5560780752d1 0x5560783f81df 0x556078475a04 0x5560783fecd9 0x5560781c5a60 0x556077f12c5c
addr2line -C -e /usr/lib/libc.so.6 0x7fe08cc1830f 0x7fe08cc183c00
addr2line -C -e multidomain 0x556077f11af4

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 3 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
) [0x56429fd9bae7]: mfem::mfem_error(char const*)
1) [0x56429fecf429]: mfem::Array<mfem::Element*>::operator[](int) const
2) [0x56429fed02d1]: mfem::Mesh::GetFaceGeometry(int) const
3) [0x5642a02531df]: mfem::FiniteElementSpace::GetFaceDofs(int, mfem::Array<int>&, int) const
4) [0x5642a02d0a04]: mfem::ParFiniteElementSpace::GetFaceDofs(int, mfem::Array<int>&, int) const
5) [0x5642a0259cd9]: mfem::FiniteElementSpace::GetFaceVDofs(int, mfem::Array<int>&) const
6) [0x5642a0020a60]: mfem::ParSubMesh::Transfer(mfem::ParGridFunction const&, mfem::ParGridFunction&)
```

Reply:
Thanks, that should be an easy fix. I'm adding L2 support right now and guard the calls better.

Reply:
Please try the latest commit ;)

Reply:
Ok, this is not so simple fix. The transfer doesn't look correct on >4 ranks!

Reply:
The issue here is really a conceptual-bug. The mesh distributes perfectly fine on 4 or 6 ranks. See this screenshot. The orange line is the topological attribute for the inner cylinder distributed on 4 ranks.
<img width="512" alt="Screen Shot 2022-04-28 at 8 49 28 AM" src="https://user-images.githubusercontent.com/5412886/165793095-63176671-58e7-41b6-a9f2-c64dc1b4d8ab.png">
Now on 5 ranks it looks like this.
<img width="512" alt="Screen Shot 2022-04-28 at 8 52 21 AM" src="https://user-images.githubusercontent.com/5412886/165793995-29d41616-73d6-41cf-a3a6-0ccfeab68e7b.png">
Suppose now, we split the domains using the attributes. The cylinder on attribute 1 and the outer box without the cylinder on attribute 2. We keep the distribution of elements on each processor the same.

Now I want to transfer a solution from the outer box to the cylinder surface. From this picture, it is clear that on rank 2, we need access to the faces marked with red dots in the second picture. They don't reside on the same rank anymore and therefore are not residing in the map. This is currently not handled.

Two solution ideas.

1. Introduce constraints into the partitioning. "Clamp" interface boundaries ALWAYS to the same processor. Multiple advantages like no-communication on solution transfer, no duplicated boundary elements etc. but no clear way how to realize this.
2. Introduce communication if this scenario occurs.

Reply:
I agree with all of your points :)

Reply:
I opted for 2 in my recent commit. Please try to break it :)

Reply:
The fix works great so far! :)

I dug a bit into the implementation and have a follow up question. It is intended that NC information is not propagated to the subdomains, i.e. nonconforming subdomains are not supported, correct? Otherwise feeding an NC serial mesh crashes with:

```
Verification failed: (faces_info[gf].Elem2No < 0) is false:
 --> Invalid mesh topology.  Interior edge found between 2D elements 18, 35 and 36.
 ... in function: void mfem::Mesh::AddSegmentFaceElement(int, int, int, int, int)
 ... in file: mesh/mesh.cpp:6456

Backtrace:
0) [0x55a61090d6a7]: mfem::mfem_error(char const*)
1) [0x55a610a2082b]: mfem::Mesh::AddSegmentFaceElement(int, int, int, int, int)
2) [0x55a610a006ec]: mfem::Mesh::GenerateFaces()
3) [0x55a610a0a8be]: mfem::Mesh::FinalizeTopology(bool)
4) [0x55a610b93161]: mfem::ParSubMesh::ParSubMesh(mfem::ParMesh const&, mfem::SubMesh::From, mfem::Array<int>&)
5) [0x55a610b92c4f]: mfem::ParSubMesh::CreateFromBoundary(mfem::ParMesh const&, mfem::Array<int>&)
6) [0x55a6108de587]: main
7) [0x7f00041c130f]: __libc_start_call_main
8) [0x7f00041c13c0]: __libc_start_main
9) [0x55a6108ddb34]: _start

Lookup backtrace source lines:
addr2line -C -e multidomain 0x55a61090d6a7 0x55a610a2082b 0x55a610a006ec 0x55a610a0a8be 0x55a610b93161 0x55a610b92c4f 0x55a6108de587
addr2line -C -e /usr/lib/libc.so.6 0x7f00041c130f 0x7f00041c13c0
addr2line -C -e multidomain 0x55a6108ddb34
```

If it is not intended to use NC meshes, could we maybe add an explicit error message to the (Par)SubMesh ctors to communicate the problem to an user?

Reply:
Yes non conforming meshes are not in scope for the first iteration of this feature. I'll add a specialized constructor for `NCMesh` that gives an appropriate error message. Thanks for reminding me.

Reply:
Great. Thanks for taking the time!